third_party/xla_client/stablehlo_helper.cc

```diff
@@ -0,0 +1,44 @@
+#include "third_party/xla_client/stablehlo_helper.h"
+
+#include <iostream>
+
+#include "mlir/IR/Verifier.h"       // from @llvm-project
+#include "mlir/Pass/PassManager.h"  // from @llvm-project
+#include "tensorflow/compiler/xla/mlir_hlo/mhlo/transforms/passes.h"
+#include "tensorflow/compiler/xla/translate/hlo_to_mhlo/hlo_to_mlir_hlo.h"
+#include "tensorflow/compiler/xla/translate/mhlo_to_hlo/mlir_hlo_to_hlo.h"
+#include "third_party/xla_client/debug_macros.h"
+#include "third_party/xla_client/xla_util.h"
+
+namespace xla {
+
+static void hlo_mhlo_helper(const HloModuleProto* proto,
+                            mlir::ModuleOp* mlir_module) {
+  auto status = ConvertHloToMlirHlo(*mlir_module, proto,
+                                    /*import_all_computations*/ false);
+  XLA_CHECK(status.ok()) << "HLO -> MHLO conversion failed.";
+  XLA_CHECK(mlir::verify(*mlir_module).succeeded())
+      << "MHLO from HLO->MHLO is not leagal.";
+}
+
+static void mhlo_stablehlo_helper(mlir::ModuleOp* mlir_module,
+                                  mlir::MLIRContext* context) {
+  mlir::PassManager pm(context);
+  pm.addPass(mlir::mhlo::createHloLegalizeToStablehloPass());
+  XLA_CHECK(mlir::succeeded(pm.run(*mlir_module)))
+      << "MHLO -> StableHLO conversion failed.";
+}
+
+std::string hlo_to_stablehlo_str(const HloModuleProto* proto) {
+  mlir::MLIRContext context;
+  mlir::ModuleOp mlir_module =
+      mlir::ModuleOp::create(mlir::UnknownLoc::get(&context));
+  hlo_mhlo_helper(proto, &mlir_module);
+  mhlo_stablehlo_helper(&mlir_module, &context);
+  std::string txt_mlir_module;
+  llvm::raw_string_ostream os{txt_mlir_module};
+  mlir_module.print(os);
```

Comment:
LGTM. Future enhancement ideas:

1. If the goal is for this to be used anywhere in production (i.e. dump a thing we can compile later) consider adding an option to dump as bytecode or StableHLO portabl artifact.
2. If the HLO has any interesting debug information, it will be thrown away here since prettyprint (the default) does not include file/line/col info. If preserving is desired, can add [enableDebugInfo](https://github.com/openxla/xla/blob/98f817b8d377f86e897cfd50b708412a2233ded1/xla/service/dump.cc#L620), but might not be necessary since I'm not sure if pytorch/xla includes any debug info currently (could be an interesting experiment).